### PR TITLE
Add `FeaturePyramidBackbone` and port weights from `timm` for `ResNetBackbone`

### PR DESCRIPTION
**Pull request by james77777778**
This PR introduces `FeaturePyramidBackbone`, a wrapper for `Backbone` that adds `pyramid_outputs` property.
If a vision backbone supports feature pyramids, it should subclass `FeaturePyramidBackbone`.

I modified `ResNetBackbone` by subclassing `FeaturePyramidBackbone` to include feature pyramid information.

Also, `head_dtype` is added in `ResNetImageClassifier`.

@divyashreepathihalli @mattdangerw @SamanehSaadat 

EDITED:
See https://github.com/keras-team/keras-nlp/pull/1769#issuecomment-2287967827 for updates

### CODE REVIEW

**Comment by mattdangerw** [comment]
@james77777778 I was thinking we would make `FeaturePyramidBackbone` a simple subclass of `Backbone` for most CV backbones.

So...
- `Backbone` is basically just a functional model with `from_preset`.
- `FeaturePyramidBackbone` extends `Backbone` with extra `pyramid_outputs`.
- `ResNetBackbone` extends `FeaturePyramidBackbone` directly.

The goal is to keep the `Backbone` base class as clean as we can, now that we are venturing into multi modal models with a lot of different overall patterns. `dir(bert_backbone)` shouldn't have feature pyramid stuff in it. If we could move the `token_embedding` off the backbone and into a `TextBackbone` or similar without breaking compat we probably would too.

Most CV models like ResNet, DenseNet, EfficientNet, etc, will be `FeaturePyramidBackbone`s. But something like ViT can just subclass `Backbone` directly without needing the feature pyramid outputs.

WDYT?

**Comment by james77777778** [comment]
@mattdangerw @divyashreepathihalli 
Got it. That makes sense!

I have updated the PR to make compatible with `timm`. Additionally, the conversion logic has been added, similar to how it’s done in `transformers`.

You can verify the numerical differences using the following script:

```python
import numpy as np
import requests
import timm  # pip install -U timm
import torch
from keras import ops
from PIL import Image

from keras_nlp.src.models.resnet.resnet_backbone import ResNetBackbone

url = "https://storage.googleapis.com/tensorflow/keras-applications/tests/elephant.jpg"
image = Image.open(requests.get(url, stream=True).raw)

# V1: resnet18.a1_in1k, resnet26.bt_in1k, resnet34.a1_in1k, resnet50.a1_in1k, resnet101.a1h_in1k, resnet152.a1h_in1k
# V2: resnetv2_50.a1h_in1k, resnetv2_101.a1h_in1k
model_id = "resnetv2_101.a1h_in1k"

timm_model = timm.create_model(model_id, pretrained=True).eval()
keras_model = ResNetBackbone.from_preset(
    f"hf://timm/{model_id}", load_weights=True
)

# Run inference
timm_transform = timm.data.create_transform(
    **timm.data.resolve_data_config(timm_model.pretrained_cfg)
)
timm_image = timm_transform(image)
timm_output = timm_model.forward_features(timm_image.unsqueeze(0))
timm_output = torch.mean(timm_output, dim=(2, 3))
timm_output = timm_output.detach().cpu().numpy()

keras_image = timm_transform.transforms[0](image)  # Resize
keras_image = timm_transform.transforms[1](keras_image)  # CenterCrop
keras_image = ops.convert_to_tensor(np.array(keras_image))
keras_output = keras_model(ops.expand_dims(keras_image, axis=0))
keras_output = ops.convert_to_numpy(keras_output)

np.testing.assert_allclose(keras_output, timm_output, atol=1e-5)
print("All passed!")

```
All compatible resnet variants passed the tests

Supported:
- V1: resnet18.a1_in1k, resnet26.bt_in1k, resnet34.a1_in1k, resnet50.a1_in1k, resnet101.a1h_in1k, resnet152.a1h_in1k
- V2: resnetv2_50.a1h_in1k, resnetv2_101.a1h_in1k

File: keras_nlp/src/utils/preset_utils.py [hunk 1]
```diff
@@ -50,6 +50,7 @@
 KAGGLE_PREFIX = "kaggle://"
 GS_PREFIX = "gs://"
 HF_PREFIX = "hf://"
+TIMM_PREFIX = "hf://timm"
```
Comment (mattdangerw) [inline keras_nlp/src/utils/preset_utils.py:53]: I think we can merge like this for now to keep things moving, but we should probably allow converting timm models outside of the timm "org" on huggingface. We might need to start parsing the `architecture`/`architectures` values transformers/timm config.json. At least that seems like the best place to infer where things are loading from.

**Review by mattdangerw — COMMENTED** [review]
Awesome great work! Still a few comments to resolve, but very cool we are building timm conversion into the library.

**Comment by divyashreepathihalli** [comment]
Looks really good! Its awesome to have built in timm conversion! LGTM!

File: keras_nlp/src/models/feature_pyramid_backbone.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import keras
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.models.backbone import Backbone
+
+
+@keras_nlp_export("keras_nlp.models.FeaturePyramidBackbone")
+class FeaturePyramidBackbone(Backbone):
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/feature_pyramid_backbone.py:21]: Added.

```python
    """A backbone with feature pyramid outputs.

    `FeaturePyramidBackbone` extends `Backbone` with a single `pyramid_outputs`
    property for accessing the feature pyramid outputs of the model. Subclassers
    should set the `pyramid_outputs` property during the model constructor.

    Example:

    ```python
    input_data = np.random.uniform(0, 255, size=(2, 224, 224, 3))

    # Convert to feature pyramid output format using ResNet.
    backbone = ResNetBackbone.from_preset("resnet50")
    model = keras.Model(
        inputs=backbone.inputs, outputs=backbone.pyramid_outputs
    )
    model(input_data)  # A dict containing the keys ["P2", "P3", "P4", "P5"]
    ```
    """
```

File: keras_nlp/src/models/resnet/resnet_feature_pyramid_backbone.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
+# Copyright 2024 The KerasNLP Authors
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_feature_pyramid_backbone.py]: Yes!

File: keras_nlp/src/models/resnet/resnet_image_classifier.py [hunk 4]
```diff
@@ -28,6 +28,8 @@ class ResNetImageClassifier(ImageClassifier):
         activation: `None`, str or callable. The activation function to use on
             the `Dense` layer. Set `activation=None` to return the output
             logits. Defaults to `"softmax"`.
+        head_dtype: `None` or str or `keras.mixed_precision.DTypePolicy`. The
+            dtype to use for the head's computations and weights.
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/models/resnet/resnet_image_classifier.py]: Fixed.

File: keras_nlp/src/utils/timm/safetensor_utils.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Copyright 2024 The KerasNLP Authors
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_nlp/src/utils/timm/safetensor_utils.py]: I have added a argument `preset` to make compatible with `timm`.
Now we use the same file.

File: keras_nlp/src/utils/preset_utils.py (hunk 1 above)
Comment (james77777778) [reply keras_nlp/src/utils/preset_utils.py]: I updated the code to parse these values in order to determine the format.
It's a bit fragile. We might need better parsing in the future.

**Review by divyashreepathihalli — APPROVED** [review]

File: keras_nlp/src/utils/preset_utils.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/src/utils/preset_utils.py]: Let's merge for now and update later!

**Comment by mattdangerw** [comment]
Thanks!

**Pull request merged**